### PR TITLE
Adding Header SpanDecorators to RestTemplate

### PR DESCRIPTION
**Pull request by ledor473**
Similar to https://github.com/opentracing-contrib/java-web-servlet-filter/pull/49 but for RestTemplate.

There is 2 more PR related to this

cc @jfbreault

### CODE REVIEW

File: opentracing-spring-web/pom.xml [hunk 1]
```diff
@@ -37,5 +37,16 @@
...
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
```
Comment (ledor473) [inline opentracing-spring-web/pom.xml]: This will be removed by another PR that is coming in

File: opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/BaseRestTemplateSpanDecorator.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package io.opentracing.contrib.spring.web.client.decorator;
+
+import io.opentracing.Span;
+import io.opentracing.contrib.spring.web.client.RestTemplateSpanDecorator;
+import org.springframework.http.HttpRequest;
+import org.springframework.http.client.ClientHttpResponse;
+
+public class BaseRestTemplateSpanDecorator implements RestTemplateSpanDecorator {
```
Comment (pavolloffay) [inline opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/BaseRestTemplateSpanDecorator.java]: This can be achieved by standard spring filter/interceptor and getting active span. I would rather keep the APIs provided by instrumentations small.

File: opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateMDCSpanDecorator.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package io.opentracing.contrib.spring.web.client.decorator;
+
+import io.opentracing.Span;
+import io.opentracing.contrib.mdc.MDCSpanDecorator;
```
Comment (pavolloffay) [inline opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateMDCSpanDecorator.java]: I think you forgot to add a dependency on this jar.

File: opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/BaseRestTemplateSpanDecorator.java (hunk 2 above)
Comment (ledor473) [reply opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/BaseRestTemplateSpanDecorator.java]: I'm not sure to follow, but if that helps the `ResttemplateSpanDecorator` interface is not new to this pull request, it was added here: https://github.com/opentracing-contrib/java-spring-web/blob/55cc1d7c9e8dc691d327022730e37f300e61e291/opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/RestTemplateSpanDecorator.java

File: opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateMDCSpanDecorator.java (hunk 3 above)
Comment (ledor473) [reply opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateMDCSpanDecorator.java]: It would be added once the java-web-servlet PR is merged: https://github.com/opentracing-contrib/java-web-servlet-filter/blob/0ad6fddb87bbfb820edf1caaa9242832f7088377/opentracing-web-servlet-filter/src/main/java/io/opentracing/contrib/mdc/MDCSpanDecorator.java

**Comment by ledor473** [comment]
@pavolloffay regarding this:
> As I mentioned in the previous comment I don't like adding new decorators as the same can be achieved by standard filters/interceptors and active span.
Should we get rid of the `ResttemplateSpanDecorator` class in that case?

**Comment by ledor473** [comment]
@pavolloffay I've remove all MDC from this merge request, added the javadoc and fix the naming of various fields.
The current conflict is because this branch is meant to be merge after 0.3.4.

I will port it back for 1.x.x right after it's merge in 0.x.x

**Comment by pavolloffay** [comment]
@ledor473 could you please resolve conflicts?

**Comment by ledor473** [comment]
@pavolloffay can you create a branch `release-0.x` like this one https://github.com/opentracing-contrib/java-spring-cloud/tree/release-0.1

There's no conflict if I compare on this tag, but I cannot send a PR to a tag: https://github.com/opentracing-contrib/java-spring-web/tree/0.3.4

File: opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateHeaderSpanDecorator.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import org.springframework.http.HttpRequest;
+import org.springframework.http.client.ClientHttpResponse;
+
+/**
+ * RestTemplateHeaderSpanDecorator will decorate the span based on HTTP headers of the request.
+ * Prior to the request, the headers are compared to the {@link #allowedHeaders} list, if it's part
+ * of the provided list, they will be added as {@link StringTag}.
+ * The tag format will be a concatenation of {@link #prefix} and {@link HeaderEntry#tag}
+ */
+public class RestTemplateHeaderSpanDecorator  implements RestTemplateSpanDecorator, Ordered {
```
[the remark this answers is not in the document]
Comment (ledor473) [reply opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateHeaderSpanDecorator.java]: It's probably not mandatory but it #85 allow people that are using the library to easily include there own SpanDecorator (see the usage of `ObjectProvider`) and it allows us to sort all of them in the autoconfigure like this: https://github.com/opentracing-contrib/java-spring-web/pull/85/files#diff-64700c454a15ffea7ba50cd59314a697R64

If we dont put any number, the default value is the following which makes it harder to run another decorator before/after this one:
>Any object that does not provide its own order value is implicitly assigned a value of Ordered.LOWEST_PRECEDENCE, thus ending up at the end of a sorted collection in arbitrary order with respect to other objects with the same order value.

[the remark this answers is not in the document]
Comment (pavolloffay) [reply opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateHeaderSpanDecorator.java]: What is the use case for sorting span decorators?

[the remark this answers is not in the document]
Comment (ledor473) [reply opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateHeaderSpanDecorator.java]: I'd say it could be useful if 2 decorators set the same tag, with that configuration you can decide which one will set it last.

From what I see we use it to make the MDC decorator run before the Header decorator.

It could be removed from the opensource version of it

[the remark this answers is not in the document]
Comment (pavolloffay) [reply opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateHeaderSpanDecorator.java]: Sorry, it is still not sure why the order is necessary. 

The decorator operates on Span and request. OT spec does not say how should implementation handle multiple tags with the same key.

[the remark this answers is not in the document]
Comment (pavolloffay) [reply opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateHeaderSpanDecorator.java]: I would rather remove it and add it later if needed. Could you please also rebase?

[the remark this answers is not in the document]
Comment (williamokano) [reply opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateHeaderSpanDecorator.java]: I'm so sorry for this being not related to the PR, but I got really curious. You said that `It could be removed from the opensource version of it`, so I got this question: does opentracing has a paid version? If so, could you provide me some links and benefits for using a paid version? Thanks and again, sorry for putting this inside a PR, I was just "reviewing" the code and this really got my attention.

[the remark this answers is not in the document]
Comment (pavolloffay) [reply opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/decorator/RestTemplateHeaderSpanDecorator.java]: There is no paid version. However, companies can have private forks with add functionality which is related yo their use cases and do not widely apply to everybody.
